SimpleArangoRepository.save() fails with Error: 1466 - must not specify _key for this collection

From digging into this, the problem is exposed in the following scenario:
* running Arango in a cluster, not single-instance
* The underlying Arango collection uses a custom shard key (which is itself necessary to allow configuration of a custom unique hash index - see #173)
* ArangoDB version is >= 3.4.0

I am running with:
-  `arangodb-spring-data`, version 3.2.3
-  `arangodb-java-driver`, version 6.6.2
-  ArangoDB itself, version 3.6.3.

Note that our domain object has the following annotations (I'm using Lombok for a couple of these):
```java
@EqualsAndHashCode
@Data
@Document(value = "myCollectionName", shardKeys = "myCustomId")
@HashIndex(fields = { "myCustomId" }, unique = true)
public class MyClassName implements Serializable {

    private static final long serialVersionUID = 1L;

    @ArangoId
    private String id;

    @Id
    private String key;

    private long documentVersion = 1L;
    private long myCustomId;
    ... other fields ...
}
```

As per the code in https://github.com/arangodb/spring-data/blob/master/src/main/java/com/arangodb/springframework/repository/SimpleArangoRepository.java#L86-L90, the `save()` method end up calling repsert().  It fails every time with a stack trace like this:

```
org.springframework.dao.InvalidDataAccessApiUsageException: Response: 400, Error: 1466 - must not specify _key for this collection; nested exception is com.arangodb.ArangoDBException: Response: 400, Error: 1466 - must not specify _key for this collection
    at com.arangodb.springframework.core.util.ArangoExceptionTranslator.translateExceptionIfPossible(ArangoExceptionTranslator.java:57)
    at com.arangodb.springframework.core.template.ArangoTemplate.insert(ArangoTemplate.java:583)
    at com.arangodb.springframework.core.template.ArangoTemplate.repsert(ArangoTemplate.java:688)
    at com.arangodb.springframework.repository.SimpleArangoRepository.save(SimpleArangoRepository.java:89)
    ... <truncated for brevity>
Caused by: com.arangodb.ArangoDBException: Response: 400, Error: 1466 - must not specify _key for this collection
    at com.arangodb.internal.util.ResponseUtils.checkError(ResponseUtils.java:52)
    at com.arangodb.internal.velocystream.VstCommunication.checkError(VstCommunication.java:141)
    at com.arangodb.internal.velocystream.VstCommunicationSync.execute(VstCommunicationSync.java:126)
    at com.arangodb.internal.velocystream.VstCommunicationSync.execute(VstCommunicationSync.java:40)
    at com.arangodb.internal.velocystream.VstCommunication.execute(VstCommunication.java:124)
    at com.arangodb.internal.velocystream.VstProtocol.execute(VstProtocol.java:46)
    at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:71)
    at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:61)
    at com.arangodb.internal.ArangoCollectionImpl.insertDocument(ArangoCollectionImpl.java:56)
    at com.arangodb.springframework.core.template.ArangoTemplate.insert(ArangoTemplate.java:581)
    ... 39 common frames omitted
```

The documentation at https://www.arangodb.com/docs/stable/release-notes-new-features34.html specifically says:
>Please note that in a cluster setup the Repsert operation requires the collection to be sharded by _key.

So from where I'm standing, this looks like a straight-up bug in spring-data-arangodb.

--- a/src/main/java/com/arangodb/springframework/core/ArangoOperations.java
+++ b/src/main/java/com/arangodb/springframework/core/ArangoOperations.java
@@ -486,7 +486,6 @@ public interface ArangoOperations {
 	 *            A representation of a single document
 	 * @throws DataAccessException
 	 * @since ArangoDB 3.4
-	 *
 	 */
 	<T> void repsert(T value) throws DataAccessException;
 
@@ -494,8 +493,10 @@ public interface ArangoOperations {
 	 * Creates new documents from the given documents, unless there already exists. In that case it replaces the
 	 * documents.
 	 *
-	 * @param values      documents to save
-	 * @param entityClass The entity class which represents the collection
+	 * @param values
+	 *            A List of documents
+	 * @param entityClass
+	 *            The entity class which represents the collection
 	 * @throws DataAccessException
 	 * @since ArangoDB 3.4
 	 */

--- a/src/main/java/com/arangodb/springframework/core/ArangoOperations.java
+++ b/src/main/java/com/arangodb/springframework/core/ArangoOperations.java
@@ -505,6 +505,8 @@ public interface ArangoOperations {
 	 * return {@param values: Iterable<T>} (same identical instances)
 	 * <p>
 	 * TODO: refactor use of generic
+	 * TODO: add Class<T> as argument, repository should know it from related entity information
+	 * 		 also in this way we keep the same fn signature as b4, so w/o breaking changes
 	 */
 	<T> void repsert(Iterable<? extends T> values) throws DataAccessException;
 

--- a/src/main/java/com/arangodb/springframework/core/ArangoOperations.java
+++ b/src/main/java/com/arangodb/springframework/core/ArangoOperations.java
@@ -487,9 +487,6 @@ public interface ArangoOperations {
 	 * @throws DataAccessException
 	 * @since ArangoDB 3.4
 	 *
-	 * TODO:
-	 * return {@param value: T} (same identical instance)
-	 *
 	 */
 	<T> void repsert(T value) throws DataAccessException;
 
@@ -497,18 +494,12 @@ public interface ArangoOperations {
 	 * Creates new documents from the given documents, unless there already exists. In that case it replaces the
 	 * documents.
 	 *
-	 * @param values A List of documents
+	 * @param values      documents to save
+	 * @param entityClass The entity class which represents the collection
 	 * @throws DataAccessException
 	 * @since ArangoDB 3.4
-	 * <p>
-	 * TODO:
-	 * return {@param values: Iterable<T>} (same identical instances)
-	 * <p>
-	 * TODO: refactor use of generic
-	 * TODO: add Class<T> as argument, repository should know it from related entity information
-	 * 		 also in this way we keep the same fn signature as b4, so w/o breaking changes
 	 */
-	<T> void repsert(Iterable<? extends T> values) throws DataAccessException;
+	<T> void repsert(Iterable<? extends T> values, Class<T> entityClass) throws DataAccessException;
 
 	/**
 	 * Checks whether the document exists by reading a single document head

--- a/src/main/java/com/arangodb/springframework/core/ArangoOperations.java
+++ b/src/main/java/com/arangodb/springframework/core/ArangoOperations.java
@@ -20,25 +20,18 @@
 
 package com.arangodb.springframework.core;
 
-import java.util.Map;
-import java.util.Optional;
-
-import org.springframework.dao.DataAccessException;
-
 import com.arangodb.ArangoCursor;
 import com.arangodb.ArangoDB;
 import com.arangodb.entity.ArangoDBVersion;
 import com.arangodb.entity.DocumentEntity;
 import com.arangodb.entity.MultiDocumentEntity;
 import com.arangodb.entity.UserEntity;
-import com.arangodb.model.AqlQueryOptions;
-import com.arangodb.model.CollectionCreateOptions;
-import com.arangodb.model.DocumentCreateOptions;
-import com.arangodb.model.DocumentDeleteOptions;
-import com.arangodb.model.DocumentReadOptions;
-import com.arangodb.model.DocumentReplaceOptions;
-import com.arangodb.model.DocumentUpdateOptions;
+import com.arangodb.model.*;
 import com.arangodb.springframework.core.convert.ArangoConverter;
+import org.springframework.dao.DataAccessException;
+
+import java.util.Map;
+import java.util.Optional;
 
 /**
  * Interface that specifies a basic set of ArangoDB operations.
@@ -493,6 +486,10 @@ public interface ArangoOperations {
 	 *            A representation of a single document
 	 * @throws DataAccessException
 	 * @since ArangoDB 3.4
+	 *
+	 * TODO:
+	 * return {@param value: T} (same identical instance)
+	 *
 	 */
 	<T> void repsert(T value) throws DataAccessException;
 
@@ -500,14 +497,16 @@ public interface ArangoOperations {
 	 * Creates new documents from the given documents, unless there already exists. In that case it replaces the
 	 * documents.
 	 *
-	 * @param value
-	 *            A List of documents
-	 * @param entityClass
-	 *            The entity class which represents the collection
+	 * @param values A List of documents
 	 * @throws DataAccessException
 	 * @since ArangoDB 3.4
+	 * <p>
+	 * TODO:
+	 * return {@param values: Iterable<T>} (same identical instances)
+	 * <p>
+	 * TODO: refactor use of generic
 	 */
-	<T> void repsert(Iterable<T> value, Class<T> entityClass) throws DataAccessException;
+	<T> void repsert(Iterable<? extends T> values) throws DataAccessException;
 
 	/**
 	 * Checks whether the document exists by reading a single document head

--- a/src/main/java/com/arangodb/springframework/core/template/ArangoTemplate.java
+++ b/src/main/java/com/arangodb/springframework/core/template/ArangoTemplate.java
@@ -578,6 +578,17 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 		return insert(collectionName, value, new DocumentCreateOptions());
 	}
 
+	private static final String REPSERT_QUERY_BODY =
+			"UPSERT { _key: doc._key } " +
+					"INSERT doc._key == null ? UNSET(doc, \"_key\") : doc " +
+					"REPLACE doc " +
+					"IN @@col " +
+					"OPTIONS { ignoreRevs: false } " +
+					"RETURN NEW";
+
+	private static final String REPSERT_QUERY = "LET doc = @doc " + REPSERT_QUERY_BODY;
+	private static final String REPSERT_MANY_QUERY = "FOR doc IN @docs " + REPSERT_QUERY_BODY;
+
 	private Object getDocumentKey(final ArangoPersistentEntity<?> entity, final Object value) {
 		Object id = entity.getIdentifierAccessor(value).getIdentifier();
 		if (id == null) {
@@ -591,15 +602,22 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 
 	@Override
 	public <T> void upsert(final T value, final UpsertStrategy strategy) throws DataAccessException {
+		if (UpsertStrategy.REPLACE.equals(strategy)) {
+			repsert(value);
+			return;
+		}
+
+		// TODO: implement also UPSERT-UPDATE via AQL
+
 		final Class<? extends Object> entityClass = value.getClass();
 		final ArangoPersistentEntity<?> entity = getConverter().getMappingContext().getPersistentEntity(entityClass);
 
 		final Object id = getDocumentKey(entity, value);
 		if (id != null && (!(value instanceof Persistable) || !Persistable.class.cast(value).isNew())) {
 			switch (strategy) {
-			case UPDATE:
-				update(id.toString(), value);
-				break;
+				case UPDATE:
+					update(id.toString(), value);
+					break;
 			case REPLACE:
 			default:
 				replace(id.toString(), value);
@@ -617,6 +635,14 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 		if (!first.isPresent()) {
 			return;
 		}
+
+		if (UpsertStrategy.REPLACE.equals(strategy)) {
+			repsert(value);
+			return;
+		}
+
+		// TODO: implement also UPSERT-UPDATE via AQL
+
 		final Class<T> entityClass = (Class<T>) first.get().getClass();
 		final ArangoPersistentEntity<?> entity = getConverter().getMappingContext().getPersistentEntity(entityClass);
 
@@ -648,19 +674,72 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 
 	@Override
 	public <T> void repsert(final T value) throws DataAccessException {
-		insert(value, new DocumentCreateOptions().overwrite(true));
+		@SuppressWarnings("unchecked") final Class<T> clazz = (Class<T>) value.getClass();
+		final String collectionName = _collection(clazz).name();
+
+		potentiallyEmitEvent(new BeforeSaveEvent<>(value));
+
+		final T result;
+		try {
+			result = query(
+					REPSERT_QUERY,
+					new MapBuilder()
+							.put("@col", collectionName)
+							.put("doc", value)
+							.get(),
+					clazz
+			).first();
+		} catch (final ArangoDBException e) {
+			throw exceptionTranslator.translateExceptionIfPossible(e);
+		}
+
+		updateDBFieldsFromObject(value, result);
+		potentiallyEmitEvent(new AfterSaveEvent<>(result));
 	}
 
 	@Override
-	public <T> void repsert(final Iterable<T> value, final Class<T> entityClass) throws DataAccessException {
-		insert(value, entityClass, new DocumentCreateOptions().overwrite(true));
+	public <T> void repsert(final Iterable<? extends T> values) throws DataAccessException {
+		if (!values.iterator().hasNext()) {
+			return;
+		}
+
+		@SuppressWarnings("unchecked") final Class<? extends T> clazz = (Class<? extends T>) values.iterator().next().getClass();
+		final String collectionName = _collection(clazz).name();
+
+		potentiallyEmitBeforeSaveEvent(values);
+
+		final Iterable<? extends T> result;
+		try {
+			result = query(
+					REPSERT_MANY_QUERY,
+					new MapBuilder()
+							.put("@col", collectionName)
+							.put("docs", values)
+							.get(),
+					clazz
+			).asListRemaining();
+		} catch (final ArangoDBException e) {
+			throw translateExceptionIfPossible(e);
+		}
+
+		updateDBFieldsFromDocumentEntities(values, result);
+		result.forEach(it -> potentiallyEmitEvent(new AfterSaveEvent<>(it)));
+	}
+
+	private void updateDBFieldsFromDocumentEntities(final Iterable<?> values, final Iterable<?> res) {
+		// FIXME: res could contain errors
+		final Iterator<?> valueIterator = values.iterator();
+		final Iterator<?> resIterator = res.iterator();
+		while (valueIterator.hasNext() && resIterator.hasNext()) {
+			updateDBFieldsFromObject(valueIterator.next(), resIterator.next());
+		}
 	}
 
 	private <T> void updateDBFields(final Iterable<T> values, final MultiDocumentEntity<? extends DocumentEntity> res) {
 		final Iterator<T> valueIterator = values.iterator();
 		if (res.getErrors().isEmpty()) {
 			final Iterator<? extends DocumentEntity> documentIterator = res.getDocuments().iterator();
-			for (; valueIterator.hasNext() && documentIterator.hasNext();) {
+			for (; valueIterator.hasNext() && documentIterator.hasNext(); ) {
 				updateDBFields(valueIterator.next(), documentIterator.next());
 			}
 		} else {
@@ -686,6 +765,34 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 				.ifPresent(arangoId -> accessor.setProperty(arangoId, documentEntity.getId()));
 		entity.getRevProperty().filter(rev -> !rev.isImmutable())
 				.ifPresent(rev -> accessor.setProperty(rev, documentEntity.getRev()));
+	}
+
+	private void updateDBFieldsFromObject(final Object toModify, final Object toRead) {
+		final ArangoPersistentEntity<?> entityToRead = converter.getMappingContext().getPersistentEntity(toRead.getClass());
+		final PersistentPropertyAccessor<?> accessorToRead = entityToRead.getPropertyAccessor(toRead);
+		final ArangoPersistentProperty idPropertyToRead = entityToRead.getIdProperty();
+		final Optional<ArangoPersistentProperty> arangoIdPropertyToReadOptional = entityToRead.getArangoIdProperty();
+		final Optional<ArangoPersistentProperty> revPropertyToReadOptional = entityToRead.getRevProperty();
+
+		final ArangoPersistentEntity<?> entityToModify = converter.getMappingContext().getPersistentEntity(toModify.getClass());
+		final PersistentPropertyAccessor<?> accessorToWrite = entityToModify.getPropertyAccessor(toModify);
+		final ArangoPersistentProperty idPropertyToWrite = entityToModify.getIdProperty();
+
+		if (idPropertyToWrite != null && !idPropertyToWrite.isImmutable()) {
+			accessorToWrite.setProperty(idPropertyToWrite, accessorToRead.getProperty(idPropertyToRead));
+		}
+
+		if (arangoIdPropertyToReadOptional.isPresent()) {
+			ArangoPersistentProperty arangoIdPropertyToRead = arangoIdPropertyToReadOptional.get();
+			entityToModify.getArangoIdProperty().filter(arangoId -> !arangoId.isImmutable())
+					.ifPresent(arangoId -> accessorToWrite.setProperty(arangoId, accessorToRead.getProperty(arangoIdPropertyToRead)));
+		}
+
+		if (revPropertyToReadOptional.isPresent()) {
+			ArangoPersistentProperty revPropertyToRead = revPropertyToReadOptional.get();
+			entityToModify.getRevProperty().filter(rev -> !rev.isImmutable())
+					.ifPresent(rev -> accessorToWrite.setProperty(rev, accessorToRead.getProperty(revPropertyToRead)));
+		}
 	}
 
 	@Override

--- a/src/main/java/com/arangodb/springframework/core/template/ArangoTemplate.java
+++ b/src/main/java/com/arangodb/springframework/core/template/ArangoTemplate.java
@@ -636,14 +636,14 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 			return;
 		}
 
+		final Class<T> entityClass = (Class<T>) first.get().getClass();
 		if (UpsertStrategy.REPLACE.equals(strategy)) {
-			repsert(value);
+			repsert(value, entityClass);
 			return;
 		}
 
 		// TODO: implement also UPSERT-UPDATE via AQL
 
-		final Class<T> entityClass = (Class<T>) first.get().getClass();
 		final ArangoPersistentEntity<?> entity = getConverter().getMappingContext().getPersistentEntity(entityClass);
 
 		final Collection<T> withId = new ArrayList<>();
@@ -698,14 +698,12 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 	}
 
 	@Override
-	public <T> void repsert(final Iterable<? extends T> values) throws DataAccessException {
+	public <T> void repsert(final Iterable<? extends T> values, final Class<T> entityClass) throws DataAccessException {
 		if (!values.iterator().hasNext()) {
 			return;
 		}
 
-		@SuppressWarnings("unchecked") final Class<? extends T> clazz = (Class<? extends T>) values.iterator().next().getClass();
-		final String collectionName = _collection(clazz).name();
-
+		final String collectionName = _collection(entityClass).name();
 		potentiallyEmitBeforeSaveEvent(values);
 
 		final Iterable<? extends T> result;
@@ -716,7 +714,7 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 							.put("@col", collectionName)
 							.put("docs", values)
 							.get(),
-					clazz
+					entityClass
 			).asListRemaining();
 		} catch (final ArangoDBException e) {
 			throw translateExceptionIfPossible(e);
@@ -727,7 +725,6 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 	}
 
 	private void updateDBFieldsFromDocumentEntities(final Iterable<?> values, final Iterable<?> res) {
-		// FIXME: res could contain errors
 		final Iterator<?> valueIterator = values.iterator();
 		final Iterator<?> resIterator = res.iterator();
 		while (valueIterator.hasNext() && resIterator.hasNext()) {

--- a/src/main/java/com/arangodb/springframework/core/template/ArangoTemplate.java
+++ b/src/main/java/com/arangodb/springframework/core/template/ArangoTemplate.java
@@ -67,6 +67,17 @@ import java.util.stream.StreamSupport;
  */
 public class ArangoTemplate implements ArangoOperations, CollectionCallback, ApplicationContextAware {
 
+	private static final String REPSERT_QUERY_BODY =
+			"UPSERT { _key: doc._key } " +
+					"INSERT doc._key == null ? UNSET(doc, \"_key\") : doc " +
+					"REPLACE doc " +
+					"IN @@col " +
+					"OPTIONS { ignoreRevs: false } " +
+					"RETURN NEW";
+
+	private static final String REPSERT_QUERY = "LET doc = @doc " + REPSERT_QUERY_BODY;
+	private static final String REPSERT_MANY_QUERY = "FOR doc IN @docs " + REPSERT_QUERY_BODY;
+
 	private static final SpelExpressionParser PARSER = new SpelExpressionParser();
 
 	private volatile ArangoDBVersion version;
@@ -578,17 +589,6 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 		return insert(collectionName, value, new DocumentCreateOptions());
 	}
 
-	private static final String REPSERT_QUERY_BODY =
-			"UPSERT { _key: doc._key } " +
-					"INSERT doc._key == null ? UNSET(doc, \"_key\") : doc " +
-					"REPLACE doc " +
-					"IN @@col " +
-					"OPTIONS { ignoreRevs: false } " +
-					"RETURN NEW";
-
-	private static final String REPSERT_QUERY = "LET doc = @doc " + REPSERT_QUERY_BODY;
-	private static final String REPSERT_MANY_QUERY = "FOR doc IN @docs " + REPSERT_QUERY_BODY;
-
 	private Object getDocumentKey(final ArangoPersistentEntity<?> entity, final Object value) {
 		Object id = entity.getIdentifierAccessor(value).getIdentifier();
 		if (id == null) {
@@ -602,22 +602,15 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 
 	@Override
 	public <T> void upsert(final T value, final UpsertStrategy strategy) throws DataAccessException {
-		if (UpsertStrategy.REPLACE.equals(strategy)) {
-			repsert(value);
-			return;
-		}
-
-		// TODO: implement also UPSERT-UPDATE via AQL
-
 		final Class<? extends Object> entityClass = value.getClass();
 		final ArangoPersistentEntity<?> entity = getConverter().getMappingContext().getPersistentEntity(entityClass);
 
 		final Object id = getDocumentKey(entity, value);
 		if (id != null && (!(value instanceof Persistable) || !Persistable.class.cast(value).isNew())) {
 			switch (strategy) {
-				case UPDATE:
-					update(id.toString(), value);
-					break;
+			case UPDATE:
+				update(id.toString(), value);
+				break;
 			case REPLACE:
 			default:
 				replace(id.toString(), value);
@@ -635,15 +628,7 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 		if (!first.isPresent()) {
 			return;
 		}
-
 		final Class<T> entityClass = (Class<T>) first.get().getClass();
-		if (UpsertStrategy.REPLACE.equals(strategy)) {
-			repsert(value, entityClass);
-			return;
-		}
-
-		// TODO: implement also UPSERT-UPDATE via AQL
-
 		final ArangoPersistentEntity<?> entity = getConverter().getMappingContext().getPersistentEntity(entityClass);
 
 		final Collection<T> withId = new ArrayList<>();
@@ -720,48 +705,16 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 			throw translateExceptionIfPossible(e);
 		}
 
-		updateDBFieldsFromDocumentEntities(values, result);
+		updateDBFieldsFromObjects(values, result);
 		result.forEach(it -> potentiallyEmitEvent(new AfterSaveEvent<>(it)));
 	}
 
-	private void updateDBFieldsFromDocumentEntities(final Iterable<?> values, final Iterable<?> res) {
+	private void updateDBFieldsFromObjects(final Iterable<?> values, final Iterable<?> res) {
 		final Iterator<?> valueIterator = values.iterator();
 		final Iterator<?> resIterator = res.iterator();
 		while (valueIterator.hasNext() && resIterator.hasNext()) {
 			updateDBFieldsFromObject(valueIterator.next(), resIterator.next());
 		}
-	}
-
-	private <T> void updateDBFields(final Iterable<T> values, final MultiDocumentEntity<? extends DocumentEntity> res) {
-		final Iterator<T> valueIterator = values.iterator();
-		if (res.getErrors().isEmpty()) {
-			final Iterator<? extends DocumentEntity> documentIterator = res.getDocuments().iterator();
-			for (; valueIterator.hasNext() && documentIterator.hasNext(); ) {
-				updateDBFields(valueIterator.next(), documentIterator.next());
-			}
-		} else {
-			final Iterator<Object> documentIterator = res.getDocumentsAndErrors().iterator();
-			for (; valueIterator.hasNext() && documentIterator.hasNext();) {
-				final Object nextDoc = documentIterator.next();
-				final Object nextValue = valueIterator.next();
-				if (nextDoc instanceof DocumentEntity) {
-					updateDBFields(nextValue, (DocumentEntity) nextDoc);
-				}
-			}
-		}
-	}
-
-	private void updateDBFields(final Object value, final DocumentEntity documentEntity) {
-		final ArangoPersistentEntity<?> entity = converter.getMappingContext().getPersistentEntity(value.getClass());
-		final PersistentPropertyAccessor<?> accessor = entity.getPropertyAccessor(value);
-		final ArangoPersistentProperty idProperty = entity.getIdProperty();
-		if (idProperty != null && !idProperty.isImmutable()) {
-			accessor.setProperty(idProperty, documentEntity.getKey());
-		}
-		entity.getArangoIdProperty().filter(arangoId -> !arangoId.isImmutable())
-				.ifPresent(arangoId -> accessor.setProperty(arangoId, documentEntity.getId()));
-		entity.getRevProperty().filter(rev -> !rev.isImmutable())
-				.ifPresent(rev -> accessor.setProperty(rev, documentEntity.getRev()));
 	}
 
 	private void updateDBFieldsFromObject(final Object toModify, final Object toRead) {
@@ -790,6 +743,38 @@ public class ArangoTemplate implements ArangoOperations, CollectionCallback, App
 			entityToModify.getRevProperty().filter(rev -> !rev.isImmutable())
 					.ifPresent(rev -> accessorToWrite.setProperty(rev, accessorToRead.getProperty(revPropertyToRead)));
 		}
+	}
+
+	private <T> void updateDBFields(final Iterable<T> values, final MultiDocumentEntity<? extends DocumentEntity> res) {
+		final Iterator<T> valueIterator = values.iterator();
+		if (res.getErrors().isEmpty()) {
+			final Iterator<? extends DocumentEntity> documentIterator = res.getDocuments().iterator();
+			for (; valueIterator.hasNext() && documentIterator.hasNext();) {
+				updateDBFields(valueIterator.next(), documentIterator.next());
+			}
+		} else {
+			final Iterator<Object> documentIterator = res.getDocumentsAndErrors().iterator();
+			for (; valueIterator.hasNext() && documentIterator.hasNext();) {
+				final Object nextDoc = documentIterator.next();
+				final Object nextValue = valueIterator.next();
+				if (nextDoc instanceof DocumentEntity) {
+					updateDBFields(nextValue, (DocumentEntity) nextDoc);
+				}
+			}
+		}
+	}
+
+	private void updateDBFields(final Object value, final DocumentEntity documentEntity) {
+		final ArangoPersistentEntity<?> entity = converter.getMappingContext().getPersistentEntity(value.getClass());
+		final PersistentPropertyAccessor<?> accessor = entity.getPropertyAccessor(value);
+		final ArangoPersistentProperty idProperty = entity.getIdProperty();
+		if (idProperty != null && !idProperty.isImmutable()) {
+			accessor.setProperty(idProperty, documentEntity.getKey());
+		}
+		entity.getArangoIdProperty().filter(arangoId -> !arangoId.isImmutable())
+				.ifPresent(arangoId -> accessor.setProperty(arangoId, documentEntity.getId()));
+		entity.getRevProperty().filter(rev -> !rev.isImmutable())
+				.ifPresent(rev -> accessor.setProperty(rev, documentEntity.getRev()));
 	}
 
 	@Override

--- a/src/main/java/com/arangodb/springframework/repository/SimpleArangoRepository.java
+++ b/src/main/java/com/arangodb/springframework/repository/SimpleArangoRepository.java
@@ -63,7 +63,7 @@ public class SimpleArangoRepository<T, ID> implements ArangoRepository<T, ID> {
 	}
 
 	/**
-	 * Saves the passed entity to the database using upsert from the template
+	 * Saves the passed entity to the database using repsert from the template
 	 *
 	 * @param entity
 	 *            the entity to be saved to the database
@@ -76,7 +76,7 @@ public class SimpleArangoRepository<T, ID> implements ArangoRepository<T, ID> {
 	}
 
 	/**
-	 * Saves the given iterable of entities to the database
+	 * Saves the given iterable of entities to the database using repsert from the template
 	 *
 	 * @param entities
 	 *            the iterable of entities to be saved to the database

--- a/src/main/java/com/arangodb/springframework/repository/SimpleArangoRepository.java
+++ b/src/main/java/com/arangodb/springframework/repository/SimpleArangoRepository.java
@@ -84,7 +84,7 @@ public class SimpleArangoRepository<T, ID> implements ArangoRepository<T, ID> {
 	 */
 	@Override
 	public <S extends T> Iterable<S> saveAll(final Iterable<S> entities) {
-		arangoOperations.repsert(entities);
+		arangoOperations.repsert(entities, domainClass);
 		return entities;
 	}
 

--- a/src/main/java/com/arangodb/springframework/repository/SimpleArangoRepository.java
+++ b/src/main/java/com/arangodb/springframework/repository/SimpleArangoRepository.java
@@ -20,29 +20,18 @@
 
 package com.arangodb.springframework.repository;
 
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-import java.util.stream.StreamSupport;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-import org.springframework.data.domain.Example;
-import org.springframework.data.domain.Page;
-import org.springframework.data.domain.PageImpl;
-import org.springframework.data.domain.Pageable;
-import org.springframework.data.domain.Sort;
-import org.springframework.lang.Nullable;
-import org.springframework.stereotype.Repository;
-
 import com.arangodb.ArangoCursor;
 import com.arangodb.model.AqlQueryOptions;
 import com.arangodb.springframework.core.ArangoOperations;
-import com.arangodb.springframework.core.ArangoOperations.UpsertStrategy;
 import com.arangodb.springframework.core.mapping.ArangoMappingContext;
 import com.arangodb.springframework.core.util.AqlUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.data.domain.*;
+import org.springframework.lang.Nullable;
+import org.springframework.stereotype.Repository;
+
+import java.util.*;
 
 /**
  * The implementation of all CRUD, paging and sorting functionality in ArangoRepository from the Spring Data Commons
@@ -80,14 +69,9 @@ public class SimpleArangoRepository<T, ID> implements ArangoRepository<T, ID> {
 	 *            the entity to be saved to the database
 	 * @return the updated entity with any id/key/rev saved
 	 */
-	@SuppressWarnings("deprecation")
 	@Override
 	public <S extends T> S save(final S entity) {
-		if (arangoOperations.getVersion().getVersion().compareTo("3.4.0") < 0) {
-			arangoOperations.upsert(entity, UpsertStrategy.REPLACE);
-		} else {
-			arangoOperations.repsert(entity);
-		}
+		arangoOperations.repsert(entity);
 		return entity;
 	}
 
@@ -98,15 +82,9 @@ public class SimpleArangoRepository<T, ID> implements ArangoRepository<T, ID> {
 	 *            the iterable of entities to be saved to the database
 	 * @return the iterable of updated entities with any id/key/rev saved in each entity
 	 */
-	@SuppressWarnings("deprecation")
 	@Override
 	public <S extends T> Iterable<S> saveAll(final Iterable<S> entities) {
-		if (arangoOperations.getVersion().getVersion().compareTo("3.4.0") < 0) {
-			arangoOperations.upsert(entities, UpsertStrategy.UPDATE);
-		} else {
-			final S first = StreamSupport.stream(entities.spliterator(), false).findFirst().get();
-			arangoOperations.repsert(entities, (Class<S>) first.getClass());
-		}
+		arangoOperations.repsert(entities);
 		return entities;
 	}
 

--- a/src/test/java/com/arangodb/springframework/repository/ShardedCollectionRepositoryTest.java
+++ b/src/test/java/com/arangodb/springframework/repository/ShardedCollectionRepositoryTest.java
@@ -1,0 +1,71 @@
+/*
+ * DISCLAIMER
+ *
+ * Copyright 2016 ArangoDB GmbH, Cologne, Germany
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright holder is ArangoDB GmbH, Cologne, Germany
+ */
+
+package com.arangodb.springframework.repository;
+
+import com.arangodb.springframework.AbstractArangoTest;
+import com.arangodb.springframework.annotation.Document;
+import org.junit.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.annotation.Id;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Michele Rastelli
+ */
+public class ShardedCollectionRepositoryTest extends AbstractArangoTest {
+
+    @Autowired
+    ShardedRepository shardedRepository;
+
+    @Test
+    public void save() {
+        ShardedUser d1 = shardedRepository.save(new ShardedUser(null, "name1", "country1"));
+        d1.name = "name2";
+        ShardedUser d2 = shardedRepository.save(d1);
+        assertThat(d2.key).isEqualTo(d1.key);
+        assertThat(d2.country).isEqualTo("country1");
+        assertThat(d2.name).isEqualTo("name2");
+
+        ShardedUser d3 = shardedRepository.save(new ShardedUser(d1.key, "name3", "country1"));
+        assertThat(d3.key).isEqualTo(d1.key);
+        assertThat(d3.country).isEqualTo("country1");
+        assertThat(d3.name).isEqualTo("name3");
+    }
+}
+
+@Document(shardKeys = "country", numberOfShards = 10)
+class ShardedUser {
+    @Id
+    String key;
+    String name;
+    String country;
+
+    public ShardedUser(String key, String name, String country) {
+        this.key = key;
+        this.name = name;
+        this.country = country;
+    }
+
+}
+
+interface ShardedRepository extends ArangoRepository<ShardedUser, String> {
+}

